### DISCUSSION TITLE
Authentication fails in businesses that require registered devices.

### DISCUSSION THREAD

**Discussion by danieljaguiar** [tauri-apps/tauri#6516] [Q&A] [answered]
I am testing Tauri integration with Azure Authentication.

My current Tenant requires Devices to be Registered to authenticate in their Azure AD.

Tauri WebView works like an InPrivate/Incognito web-browser, having no access to the main web-browser user profile, thus, failing to load the required information of Device Status to finish the Azure Authentication.

![image](https://user-images.githubusercontent.com/23394221/226649536-311136ee-be42-4c77-bf1b-4d0caeed164d.png)

The error I get in the Sign In Window above is: 

```
Error Code: 53000
(...)
Device identifier: Not available
Device platform: Windows 10
Device state: Unregistered <=== CANNOT RECOGNIZE THE DEVICE IN TAURI WEBVIEW
```

It is the same error if I open Microsoft Edge as InPrivate. But in the full web-browser an extra detail is provided:

```
To access your service, app, or website, you may need to sign in to Microsoft Edge using user@tenant.com

Error Code: 53000
(...)
Device identifier: Not available
Device platform: Windows 10
Device state: Unregistered <=== CANNOT RECOGNIZE THE DEVICE IN FULL WEB-BROWSER INPRIVATE
````

So, to make Authentication work in businesses that require registered devices, the WebView needs to access, not behave like in InPrivate/Incognito but needs to "see" a few things, like, the Web-Browser profile.

I noticed the WebView in Tauri loads with the following parameters:

```
"C:\Program Files (x86)\Microsoft\EdgeWebView\Application\111.0.1661.44\msedgewebview2.exe" --embedded-browser-webview=1 --webview-exe-name=spot-app.exe --webview-exe-version=0.0.1 --user-data-dir="C:\Users\MYUSER\AppData\Local\com.microsoft.spot\EBWebView" --noerrdialogs --embedded-browser-webview-dpi-awareness=2 --disable-features=MojoIpcz,msWebOOUI,msPdfOOUI,msSmartScreenProtection --lang=en-US --mojo-named-platform-channel-pipe=60868.64160.5224727295498144568 /pfhostedapp:7130990d2f7d659ffa75b99db6a286662d14d3e7
```

Perhaps, if the parameters above were tweaked, this behavior could change?

Can Tauri WebView behave less like "InPrivate/Incognito" (e.g., loading main web-browser profile, extensions, etc.)?

_Let me know what extra details you need from my setup_

**Comment by FabianLars** [comment, marked as answer]
> Perhaps, if the parameters above were tweaked, this behavior could change?

Pretty sure we can't do that. we only invoke webview2 and it injects most of these flags itself, we can add _additional_ flags like the --disable-features stuff which comes from tauri, but we can't change already existing ones. Even if there is a way i highly doubt that we'd ever add an official api to make it possible to hijack the user's browser profiles in tauri apps...

To be honest to me this sounds like a use-case where the authentication should happen in the user's default browser, not in the app itself, would that be possible? (i genuinely don't know, i'm not familiar with azure authentication)

**Reply by danieljaguiar** [reply]
Thank you @FabianLars 

I was exploring the authentication through the default browser. Is there a way to make the default browser talk back to the Tauri  (frontend or backend)?

The Tauri Frontend (in my case React) seems to be a silo. Even if I open the Tauri Frontend direct in the default web browser(http://localhost:1420/) and set cookies there, the Tauri Frontend can't see it. 

_Opening Tauri Frontend in the default browser to test saving cookies_
![image](https://user-images.githubusercontent.com/23394221/226708123-501537d8-8f08-48f4-9a75-6b7db2f36d1b.png)

_Back to Tauri App, the cookies won't appear_
![image](https://user-images.githubusercontent.com/23394221/226708249-bce50955-b34d-4fa7-b45f-9ef1954a4c56.png)

Is there a way to make the default browser talk back to the Tauri (frontend or backend)? (Cookies, LocalStorage, etc.)

**Reply by FabianLars** [reply]
> Is there a way to make the default browser talk back to the Tauri (frontend or backend)? (Cookies, LocalStorage, etc.)

Not directly. you'd have to spawn a small localhost server, or if Azure supports it use custom uri schemes (#323)

**Reply by danieljaguiar** [reply]
This approach of "small localhost" seems to work. I am having good results with it.

In the Frontend, when I try to open a URL with multiple parameters, it fails because the 2nd parameter is lost by opening:

```
await open("https://google.com/search?q=test&param=123"); // This "&param=123" will be lost
```

In fact, if you open prompt (windows) and try to start on the URL without double quotes, you will get the error:

```
C:\Windows\System32>start https://google.com/search?q=test&param=123
'param' is not recognized as an internal or external command,
operable program or batch file.
```

The correct way would be:

```
C:\Windows\System32>start "" "https://google.com/search?q=test&param=123"
```

The first empty double quote is for Window Title, which is meaningless in this case.

Is there another way to Open url in the default browser from the Frontend?

**Reply by FabianLars** [reply]
There would be this workaround using shell.open: https://github.com/tauri-apps/tauri/issues/6473#issuecomment-1473961671 which is only necessary until we can fix and/or upgrade our dependencies

**Reply by danieljaguiar** [reply]
I was able to perform the authentication with your guidance! Thank you!
